### DISCUSSION TITLE
How do you test Helix locally?

### DISCUSSION THREAD

**Discussion by ghost** [helix-editor/helix#5506] [General] [answered]
You have a copy of the editor installed on your device. But when you work on a pull request, you modify the code and sometimes want to test it locally (manually, not unit tests). So how do you do this, what do you prefer? Is it virtualbox, docker, chroot or what? I'm afraid that if I compile the modified version of the editor, the regular version will be overwritten.

**Comment by the-mikedavis** [comment, marked as answer]
If you have a C/C++ and Rust toolchain installed along with `cargo`, you can run `cargo run` on a branch to run helix using the code on that branch. That build of helix will live in the `target/debug` directory under the helix repository while any builds of helix you installed with `cargo install --locked --path helix-term` will remain separate (under `~/.cargo/bin` if I remember correctly).

This builds helix in debug mode which is a faster build but it will be slower. You can run helix and build in release mode with `cargo run --release` for a build of helix that will be essentially the same as what you find in the [release assets](https://github.com/helix-editor/helix/releases) or from package managers

**Reply by ghost** [reply]
Oops, I forgot that if you just compile, it won't add it to the PATH, it will just save it to the target directory (in case of rust). Thank you for reminding me. :+1:

**Comment by the-mikedavis** [comment]
For another approach, a few of us use the Nix flake (the `flake.nix` file in the repository root) and the Nix package manager (and NixOS operating system) and the flake allows you to build any branch without having to clone the repository. For example you could use `nix run github:helix-editor/helix/interactive-global-search` to test #4687

I don't currently use this to review PRs (I use cargo via the Nix flake with `nix develop`) but I think you could hack something together to use `gh` to fetch the revision for a PR and then `nix run` a debug build for that revision.

**Reply by ghost** [reply]
Interesting. I wonder if this method is common enough among other open source projects, as this is the first I've heard of it.
